File: Sources/LedgelingsCore/EdgeLoop.swift
import CoreGraphics
import Foundation

/// A closed path made only of horizontal and vertical segments, walked with the
/// inside of the screens on the LEFT. Coordinates are AppKit's global ones:
/// origin bottom-left, y up.
///
/// A position on the loop is one number `t` in `0..<length`. Each segment has a
/// `rotation`: how far a sprite drawn "standing on a floor, facing right" must be
/// turned (radians, counter-clockwise) to stand on that segment with its body
/// pointing inside. Walking rightwards along a floor is rotation 0.
public struct EdgeLoop: Equatable, Sendable {
    public let vertices: [CGPoint]
    private let starts: [CGFloat]      // loop position where each segment begins
    public let length: CGFloat

    public init(vertices: [CGPoint]) {
        precondition(vertices.count >= 4, "a loop needs at least four corners")
        self.vertices = vertices
        var starts: [CGFloat] = [], total: CGFloat = 0
        for i in vertices.indices {
            starts.append(total)
            let a = vertices[i], b = vertices[(i + 1) % vertices.count]
            total += abs(b.x - a.x) + abs(b.y - a.y)
        }
        self.starts = starts
        length = total
    }

    /// The loop just inside one rectangle, starting at its bottom-left corner.
    public init(rect: CGRect) {
        self.init(vertices: [
            CGPoint(x: rect.minX, y: rect.minY), CGPoint(x: rect.maxX, y: rect.minY),
            CGPoint(x: rect.maxX, y: rect.maxY), CGPoint(x: rect.minX, y: rect.maxY),
        ])
    }

    public var segmentCount: Int { vertices.count }

    public func wrap(_ t: CGFloat) -> CGFloat {
        let m = t.truncatingRemainder(dividingBy: length)
        return m < 0 ? m + length : m
    }

    public func segment(at t: CGFloat) -> Int {
        let w = wrap(t)
        return (starts.lastIndex { $0 <= w }) ?? 0
    }

    public func length(ofSegment i: Int) -> CGFloat {
        (i + 1 < starts.count ? starts[i + 1] : length) - starts[i]
    }

    /// Unit vector along segment `i`, in walking (+t) direction.
    public func direction(ofSegment i: Int) -> CGVector {
        let a = vertices[i], b = vertices[(i + 1) % vertices.count]
        let len = max(length(ofSegment: i), .leastNonzeroMagnitude)
        return CGVector(dx: (b.x - a.x) / len, dy: (b.y - a.y) / len)
    }

    /// Unit vector pointing from segment `i` into the screen.
    public func inward(ofSegment i: Int) -> CGVector {
        let d = direction(ofSegment: i)
        return CGVector(dx: -d.dy, dy: d.dx)
    }

    public func rotation(ofSegment i: Int) -> Double {
        let d = direction(ofSegment: i)
        let angle = atan2(Double(d.dy), Double(d.dx))
        return angle < 0 ? angle + 2 * .pi : angle
    }

    public func point(at t: CGFloat) -> CGPoint {
        let i = segment(at: t), d = direction(ofSegment: i), along = wrap(t) - starts[i]
        return CGPoint(x: vertices[i].x + d.dx * along, y: vertices[i].y + d.dy * along)
    }

    public func t(onSegment i: Int, fraction: CGFloat) -> CGFloat {
        starts[i] + length(ofSegment: i) * min(max(fraction, 0), 1)
    }

    /// The loop position closest to `point`, and how far away it is.
    public func nearest(to point: CGPoint) -> (t: CGFloat, distance: CGFloat) {
        var best: (t: CGFloat, distance: CGFloat) = (0, .infinity)
        for i in vertices.indices {
            let a = vertices[i], d = direction(ofSegment: i)
            let along = min(max((point.x - a.x) * d.dx + (point.y - a.y) * d.dy, 0), length(ofSegment: i))
            let distance = hypot(a.x + d.dx * along - point.x, a.y + d.dy * along - point.y)
            if distance < best.distance { best = (starts[i] + along, distance) }
        }
        return best
    }
}

File: win/LedgelingsCore/EdgeLoop.cs
namespace Ledgelings.Core;

/// <summary>
/// A closed path made only of horizontal and vertical segments, walked with the
/// inside of the screens on the LEFT. Coordinates are global points, origin
/// bottom-left, y up.
///
/// A position on the loop is one number <c>t</c> in <c>[0, length)</c>. Each segment has a
/// <c>rotation</c>: how far a sprite drawn "standing on a floor, facing right" must be
/// turned (radians, counter-clockwise) to stand on that segment with its body
/// pointing inside. Walking rightwards along a floor is rotation 0.
/// </summary>
public sealed class EdgeLoop : IEquatable<EdgeLoop>
{
    public IReadOnlyList<Pt> Vertices { get; }
    private readonly double[] starts;      // loop position where each segment begins
    public double Length { get; }

    public EdgeLoop(IReadOnlyList<Pt> vertices)
    {
        if (vertices.Count < 4) throw new ArgumentException("a loop needs at least four corners");
        Vertices = vertices.ToArray();
        starts = new double[vertices.Count];
        double total = 0;
        for (int i = 0; i < vertices.Count; i++)
        {
            starts[i] = total;
            var a = vertices[i];
            var b = vertices[(i + 1) % vertices.Count];
            total += Math.Abs(b.X - a.X) + Math.Abs(b.Y - a.Y);
        }
        Length = total;
    }

    /// <summary>The loop just inside one rectangle, starting at its bottom-left corner.</summary>
    public EdgeLoop(Rect rect) : this(new[]
    {
        new Pt(rect.MinX, rect.MinY), new Pt(rect.MaxX, rect.MinY),
        new Pt(rect.MaxX, rect.MaxY), new Pt(rect.MinX, rect.MaxY),
    }) { }

    public int SegmentCount => Vertices.Count;

    /// <summary>The truncating remainder, made non-negative.</summary>
    public double Wrap(double t)
    {
        var m = t - Math.Truncate(t / Length) * Length;
        return m < 0 ? m + Length : m;
    }

    public int Segment(double t)
    {
        var w = Wrap(t);
        for (int i = starts.Length - 1; i >= 0; i--) if (starts[i] <= w) return i;
        return 0;
    }

    public double SegmentLength(int i) => (i + 1 < starts.Length ? starts[i + 1] : Length) - starts[i];

    /// <summary>Unit vector along segment <paramref name="i"/>, in walking (+t) direction.</summary>
    public Vec Direction(int i)
    {
        var a = Vertices[i];
        var b = Vertices[(i + 1) % Vertices.Count];
        var len = Math.Max(SegmentLength(i), double.Epsilon);
        return new Vec((b.X - a.X) / len, (b.Y - a.Y) / len);
    }

    /// <summary>Unit vector pointing from segment <paramref name="i"/> into the screen.</summary>
    public Vec Inward(int i)
    {
        var d = Direction(i);
        return new Vec(-d.Dy, d.Dx);
    }

    public double Rotation(int i)
    {
        var d = Direction(i);
        var angle = Math.Atan2(d.Dy, d.Dx);
        return angle < 0 ? angle + 2 * Math.PI : angle;
    }

    public Pt Point(double t)
    {
        var i = Segment(t);
        var d = Direction(i);
        var along = Wrap(t) - starts[i];
        return new Pt(Vertices[i].X + d.Dx * along, Vertices[i].Y + d.Dy * along);
    }

    public double T(int segment, double fraction) => starts[segment] + SegmentLength(segment) * Math.Min(Math.Max(fraction, 0), 1);

    /// <summary>The loop position closest to <paramref name="point"/>, and how far away it is.</summary>
    public (double T, double Distance) Nearest(Pt point)
    {
        (double T, double Distance) best = (0, double.PositiveInfinity);
        for (int i = 0; i < Vertices.Count; i++)
        {
            var a = Vertices[i];
            var d = Direction(i);
            var along = Math.Min(Math.Max((point.X - a.X) * d.Dx + (point.Y - a.Y) * d.Dy, 0), SegmentLength(i));
            var dx = a.X + d.Dx * along - point.X;
            var dy = a.Y + d.Dy * along - point.Y;
            var distance = Math.Sqrt(dx * dx + dy * dy);
            if (distance < best.Distance) best = (starts[i] + along, distance);
        }
        return best;
    }

    public bool Equals(EdgeLoop? other) => other is not null && Vertices.SequenceEqual(other.Vertices);
    public override bool Equals(object? obj) => Equals(obj as EdgeLoop);
    public override int GetHashCode() => Vertices.Aggregate(17, (h, v) => h * 31 + v.GetHashCode());
    public override string ToString() => "[" + string.Join(", ", Vertices) + "]";
}
